### ISSUE TITLE
Add `Group` shape to Style

### ISSUE THREAD

**Issue by keenancrane** [penrose/penrose#783]
**What's the problem?**

A longstanding deficiency in Style is that we don't provide a mechanism for defining a "group" shapes.  A group is simply an unordered collection of shapes, including other groups.  In SVG, groups are implemented as nested SVG tags.  In general, groups play several important roles in vector graphics:

- **Scene graph.** In conjunction with spatial transformations, groups enable one to define a [scene graph](https://en.wikipedia.org/wiki/Scene_graph).  A scene graph is basically a hierarchical description of how shapes in a scene are arranged.  For instance, to describe the configuration of a human character in a 3D environment, you might think of the torso as the root node; the neck and upper limbs are children of the torso; the head and lower limbs are children of the next and upper limbs; etc.  This graph can be implemented via grouping: the outermost group contains a shape for the torso, and the groups for the limbs and head/neck.  The group for a limb contains a shape for the upper part of a limb and a group for whatever is in the lower limb (including fingers and toes, etc.).  If we now associate a transformation with each group node, that node gets applied to the shapes in that node, and recursively to the child shapes.  For instance, rotating the group containing the torso rotates everything; rotating the group containing the upper limb also rotates the lower limb.  Etc.

![image](https://user-images.githubusercontent.com/3604525/148608944-25c5bf09-bca9-4a91-abbc-3d5efa64bdb9.png)

- **Clipping.** Even without transformations, groups play an important role in _clipping_.  Here, one shape is used to map out a collection of shapes.  A basic use case might be trimming objects to the bounds of the canvas.  In SVG, clipping is achieved via [the `clipPath` tag](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Clipping_and_masking).  Note that this construct does not itself rely on groups—for instance, you could apply the same clip path repeatedly to a bunch of shapes.  But the most natural thing to do is group everything that will be clipped, then apply a clip path _once_ to that group (see example below).  There is also a similar-but-different notion of [compound paths](https://www.vektorgarten.de/illustrator-compound-path-shape.html).  Such paths are used, for example, to make a hole in the letter "O".

![image](https://user-images.githubusercontent.com/3604525/148606828-a4cd2227-971d-4ecc-a07c-436859198d54.png)

- **Editability.** Like named layers (see #768), groups are also essential for making vector graphics files easily editable.  Currently Penrose just spits out a big soup of SVG shapes.  These shapes are difficult to edit in external programs like Illustrator and Inkscape (currently you have to load up the file and group elements by hand!).  Yet individual shapes very often have a natural organization.  In fact, if a `Group` shape were available in Style, it would be easy for Style programmers to define grouping based on Style rules.  Since these rules are triggered by the semantics of the Substance program, the resulting diagrams would be nicely grouped into semantically-meaningful collections of shapes.  (Or any other organization the Style writer wants to provide to a human editor---for instance, all vertices, all edges, and all triangles in a mesh could be put in separate groups).  Here's an example of an SVG that has been nicely structured by grouping, and has annotations via `id`s and `title`s.  Here I've annotated the drawing with information about spatial placement—more natural in Penrose would be to annotate it with the automatically-generated names from substance, or more descriptive names assigned by the Style programmer:

```
<svg>
   <g id="wholeDrawing">
      <rect x="654" y="174" width="208" height="208"><title>A big rectangle at the top.</title></rect>
      <g id="leftHalf">
         <rect x="535" y="427" width="138" height="138"><title>A medium rectangle at left.</title></rect>
         <g id="leftBottom">
            <rect x="459" y="598" width="85" height="85">
            <rect x="569" y="598" width="85" height="85">
         </g>
      </g>
      <g id="rightHalf">
         <rect x="838" y="427" width="138" height="138"><title>A medium rectangle at right.</title></rect>
         <g id="rightBottom">
            <rect x="883" y="598" width="85" height="85"/>
            <rect x="994" y="598" width="85" height="85"/>
         </g>
      </g>
   </g>
</svg>
```

---

**What's the proposal?**

The basic proposal is to define a new shape called `Group`.  The basic fields of a `Group` are:

1. `shapes:` — a list of shapes contained in the group.
2. `transformation:` — a spatial transformation applied to the group and everything in it.  (For now this field can be unused—we shouldn't block on implementing shapes just because we haven't completely fleshed out the design for transformations.  But a placeholder would be nice to have.)
3. `name:` — just like any other shape, a group has a name.

The list syntax for `shapes` can be whatever makes sense/is easiest to implement in Style; we just need to be able to specify a list of shapes of any length.  Just to provide a definite proposal, we could write something like

`shapes: [ x.myRect, x.myCircle, x.myPath ]`

_Optional:_ Beyond these basic fields, we could also allow arbitrary SVG attributes.  These attributes would be applied to any child that does not already define this attribute.  For instance, if we write

```
shape x.g = Group {
   shapes: [ x.myRect, x.myCircle, x.myPath ]
   stroke-width: 3.0
}
```

then the shapes `x.myRect`, `x.myCircle`, and `x.myPath` would inherit a `stroke-width` of 3.0—unless they define their own stroke width.  Basically attributes defined lower in the tree take precedence.  (And as always, fields overriden in lower blocks take precedence over fields set in earlier Style rules—group precedence and rule precedence are entirely orthogonal.) **Note:** This functionality is easily supported via SVG—for example, the following code will color the outer circle green and the two inner circles blue:

```
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 1920 1080" style="enable-background:new 0 0 1920 1080;" xml:space="preserve">
<g color="green">
	<circle cx="941.67" cy="363.67" r="156.67" fill="currentcolor"/>
	<g color="blue">
		<circle cx="818.33" cy="552" r="156.67" fill="currentcolor"/>
		<circle cx="1038.33" cy="568.67" r="156.67" fill="currentcolor"/>
	</g>
</g>
</svg>
```

![image](https://user-images.githubusercontent.com/3604525/148612227-a7327f2e-1072-4e6f-8d42-caa52136d774.png)

---

**Implementation.**

Implementation amounts to doing just a couple things:

* Define a new `Group` shape (as usual), with the properties above.
* Modify the renderer to emit grouped shapes in group order, within nested `<g>` tags.

One place where implementation gets tricky is interaction with layering.  The way this works out in basically every vector graphics setting is that all objects within a group are either above, or below, all objects within another disjoint group.  Otherwise, objects within the same group (including other groups) have some layer ordering.  For instance, if we have a group

`A = [ a1, a3, a2 ]`

and another group

`B = [ b3, b1, b2 ]`

and say that group `A` is above group `B`, then the overall ordering will end up being

`b3 < b1 < b2 < a1 < a3 < a2`.

There is no way to "interleave" ordering across groups, e.g., it's impossible for `a3` to be above `b2` *and* `a1` to be below `b3`.  The reason for this is that objects in an SVG will still get layered from top to bottom.  Since shapes within a group all appear in one contiguous chunk of tags, there's no way to interleave the layer orderings.  (Some thought still required here to totally flesh out layer order, and how to fail gracefully in a world with `Groups`…)

**Comment by keenancrane** [comment]
Some further thoughts/comments about implementation, following conversations with @cmumatt and @samestep.

### Cascading and grouping

In the proposed design, nothing prevents multiple groups from listing the same child shape.  This kind of grouping should not be allowed, because it describes a DAG rather than tree structure, which is not supported by SVG.

Instead, we can take advantage of cascading to "fail gracefully" (or, viewed a different way, to not fail at all).

In particular, let's adopt the rule that **a shape belongs to the most recent group it was named in**.  For instance, if `x.shape` is included in the `shapes` list of `group1`, then in the shapes list of `group2`, it belongs to `group2` and not `group1`.  To implement this behavior, we can store (internally) a reference to each shape's parent.  This reference will be undefined if the shape does not belong to a group.  During compilation, it gets set each time a shape appears in the `shapes` list of a group.

We could perhaps emit a warning if a shape is included in multiple groups.  But given that this behavior may even be _desirable_ for folks writing complex Style programs (rather than an unintentional blunder), it might be a good idea to make this warning optional.

### Cycle detection and prevention

This design makes it possible to create cycles in the grouping relationships.  For instance:

```
x.icon = Group {
   shapes: [ x.icon ]
}
```

Preventing cycles might be as simple as just checking, at compile time, if any shape in the list is an ancestor of the group itself; if so, refuse to add it to the list.  But there may be other good solutions, and/or pitfalls with this solution.

Here it's probably a good idea to issue a warning—but not an error.  As with inconsistent layerings, it seems much better to fail gracefully than to fail outright.  In a long cascading Style program, it can be hard to keep all the relationships in your head.  Better to get _something_ on the screen that approximately captures the program intent, rather than getting _nothing_ (which makes it hard to debug).  Breaking a few grouping relationships but preserving those that are valid also makes diagrams robust to, e.g., concatenation of multiple Style files (which we will eventually support…).

### Inherited attributes

In SVG, shapes nested inside groups automatically inherit properties of the group.  In particular, these ones:

```
 clip-path, clip-rule, color, color-interpolation, color-rendering, cursor, display, fill, fill-opacity, fill-rule, filter, mask, opacity, pointer-events, shape-rendering, stroke, stroke-dasharray, stroke-dashoffset, stroke-linecap, stroke-linejoin, stroke-miterlimit, stroke-opacity, stroke-width, transform, vector-effect, visibility
```

We should of course adopt the same semantics in Penrose.  For instance, if a `Group` shape defines a `fillColor`, and one of its child shapes does not, then the child will inherit the `fillColor`.

From a renderer implementation point of view, there is nothing to implement in order to get this behavior: just emit the properties of the group, and the children will automatically render as expected.  We do not need to, e.g., copy fields from parents into children and emit them explicitly in the SVG.

### Optimization

The most challenging aspect of implementing groups is their interaction with optimization, especially optimization of transformations.  The proposal is to adopt a "walk then run" approach to implementation:

- **first** support groups in the compiler and renderer
- **then** worry about getting groups to play well with optimization

During the "walk" stage we can also adopt a policy that still yields reasonable behavior (and has no impact on existing diagrams that do not use groups).

In more detail: consider a situation where a constraint or objective tries to use a property of a shape that is inherited from a group, rather than explicitly defined.  E.g.,
```
x.circle = Circle {
   r: 100
   center: (0,0)
}

x.group = Group {
   shapes: [ x.circle ]
   fillColor: rgba( 0, 1, 0, 1 )
}

encourage equal( x.circle.fillColor, rgba( 1, 0, 0, 1 )
```
The renderer will have no trouble with this description: even though the fill color for the circle is not defined, it will be inherited from the group.  The question is: what should the `encourage` statement do?  This is as much as question of design/semantics as it is of implementation.  There are basically two possibilities:

1. The `encourage` statement does not affect the group color.  Instead, it optimizes the field `x.circle.fillColor`.  But since `x` will inherit its render color from `x.group`, this optimization will have no effect, i.e., the optimized color is not used for anything.
2. The `encourage` statement implicitly references the group color.  I.e., since `x.circle` inherits its color from `x.group`, we should make sure this color gets optimized, and pushed closer to red than green.

The second scheme does a better job of satisfying the principle of least surprise: the user asked for the circle's color to change, and it changes!  But it's also harder to implement, because we have to modify the optimizer.  In this case, for instance, we would have to (i) check whether the circle's color had been explicitly defined, and (ii) if not, search up the parent graph to see which color it actually inherits; this color is then the one that gets optimized.

The first scheme still has semantics that can be explained to a Style programmer in simple terms: _"if you didn't define a child's attribute explicitly, you can't optimize it."_. Annoying, but not totally perplexing.  And much easier to implement: just optimize the fields that are named in the `ensure` or `encourage` statement, with no reference chasing.

TL;DR: Ultimately we want to implement (2), but it's ok to just implement (1) first.

#### Optimizing hierarchical transformations

The most challenging part will be optimizing nested transformations, since transformations get _composed_ rather than _replaced_ (unlike attributes like colors or stroke widths).  For instance, if we write

```
vec2 u = (?,?)
x.circle = Circle {
   r: 100
   center: (0,0)
   transform: Translate( u )
}

scalar theta = ?
x.group = Group {
   shapes: [ x.circle ]
   fillColor: rgba( 0, 1, 0, 1 )
   transform: Rotate( theta )
}
```

In this case, the `transform` attribute of both the circle and the group will get written to the SVG; the behavior of SVG is then to compose these transformations, so that the transformation applied to the circle is effectively `Rotate( theta ) then Translate( u )`.  A subsequent constraint like `ensure disjoint( x.circle, y.icon )` will need to account for this composite transformation in order to work properly.

The interim solution is to simply **not allow groups to have transformations**.  If the Style program specifies a `transform` field for a group, the compiler will emit a warning like,

> Sorry, transformations of Group objects are not currently supported!

This sucks, but we'll get around to implementing it soon enough.  Alternatively, we could allow groups to have transformations, but not allow optimization of transformations of grouped shapes.  To implement this behavior, we would check if a transformation passed into an `ensure` or `encourage` statement belongs to a shape with a non-null parent.  If so, we emit the warning

> Sorry, optimization of transformations inside groups are not supported!

This would be much better, since then:
   - we can still at least _draw_ groups with transformations applied to them!
   - we can also optimize the outer transformation of the group without any trouble
   - it's an easy check to implement, since we have to inspect just one field of the shape referenced from the constraint/objective

**Comment by keenancrane** [comment]
Here by the way is an example of how attributes are inherited in SVG.  Notice that the rectangle does not define a fill color, but instead inherits it from the group:

```
<svg>
<style type="text/css">
	.st0{fill:#FF6600;}
	.st1{fill:#1B1F8A;}
</style>
<g class="st1">
	<circle class="st0" cx="47.1" cy="43.5" r="31.6"/>
	<rect x="44.8" y="36.5" width="89.8" height="56.2"/>
</g>
</svg>
```

![image](https://user-images.githubusercontent.com/3604525/171032361-743d57dc-5ba4-40d3-ad1f-c0613e1b5b48.png)

**Comment by joshsunshine** [comment]
> The encourage statement does not affect the group color. Instead, it optimizes the field x.circle.fillColor. But since x will inherit its render color from x.group, this optimization will have no effect, i.e., the optimized color is not used for anything.

This doesn't make sense to me. If Penrose renders a circle with a fill color inside a group with a fill color then the circle's color will take precedence. In this SVG code: 
`<svg version="1.2" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 700 700">
<g fill="green" stroke-width="5" stroke="navy">
<circle fill="red" fill-opacity="0.5" cx="350" cy="350" r="200"></circle>
</g>`
The circle is red, not green.

**Comment by keenancrane** [comment]
@joshsunshine Yeah, there's a subtle point here:

Since, in this example, `x.shape` does not explicitly define a field `x.shape.fillColor`, it will get drawn using its parent's fill color.

The fact that we then use `x.shape.fillColor` in an `encourage` statement doesn't change this fact: only shapes that **explicitly** assign a fill color will emit this fill color to the SVG.  If _every_ shape emitted a fill color (whether specified or not), then we lose the inheritance property altogether.

Of course, this is "easy implementation path #1", not "harder implementation path #2".  In the second, harder path, the `encourage` statement would optimize the value of `x.group.fillColor`.  But even in this scenario, `x.circle` would not use its own fill color for rendering—it would still not emit a `fill-color` SVG attribute, and just inherit its group's `fill-color`.

**Comment by joshsunshine** [comment]
@keenancrane Right now, certain properties of shapes are sampled and rendered to SVG even if they are not specified in style program. Fill color is an example of such a property. Are you proposing to change that behavior for all shapes or just for shapes that are inside groups? What happens if the group itself doesn't specify a fill color (or any other property we've decided it is important to provide a default)?

**Comment by keenancrane** [comment]
@joshsunshine Yes, there will need to be changes to the renderer.  In particular, each `Shape` now internally has a `parent` attribute, which gets set to a nonempty value whenever this shape is added to a `Group`.  When we go to render an attribute of a shape (e.g., emit the `fill-color` property), we first check if it has a parent.  If so, we follow the chain of ancestors via the `parent` attribute.  If we ever find the attribute we're looking for, _we emit nothing_ for the shape itself.  Otherwise, if there is no parent, or the attribute is not found in the parental lineage, we emit the shape's own attribute.

**Comment by joshsunshine** [comment]
@keenancrane This proposal perhaps makes sense to me for values that are *not* explicitly assigned a value by style. For example, if we were to chance your example very slightly to add a `?` as so:
```
x.circle = Circle {
   r: 100
   center: (0,0)
   fillColor: ?
}

x.group = Group {
   shapes: [ x.circle ]
   fillColor: rgba( 0, 1, 0, 1 )
}

encourage equal( x.circle.fillColor, rgba( 1, 0, 0, 1 )
```
The Penrose-rendered property of the circle should be ~rgba(1,0,0,1), and the Penrose-rendered property of the group should be ~rgba( 0, 1, 0, 1 ). The SVG renderer will of course show the circle as green. Do you agree?

**Comment by keenancrane** [comment]
@joshsunshine Yeah, I would assume these semantics:

- If a constant has been assigned to a field of `x.icon`, it's of course used when drawing the shape.  It can't be optimized, because it's a constant.
- If an unknown value `?` has been assigned to a field of `x.icon`, it's also used when drawing the shape.  Its value is determined via optimization.
- If a field has not been declared for `x.icon`, the value of this field inherited from the parent and used for drawing.  In SVG terms, this means no attribute is explicitly written into the SVG file for this field.  In the first version of the optimizer, asking to optimize this undeclared value has no effect.

**Comment by joshsunshine** [comment]
I would then suggest that your original program, without the `?` should ideally behave exactly the same way as the above. Penrose should infer a new input/varying value to the optimization based on the style program.

In other words, your "run" semantics copied below, seem wrong to me.
> The encourage statement implicitly references the group color. I.e., since x.circle inherits its color from x.group, we should make sure this color gets optimized, and pushed closer to red than green.

Instead, the fillColor of the circle should be optimized.

**Comment by keenancrane** [comment]
@joshsunshine In the sentence you quote, I am considering the "harder" implementation path.  In my two previous messages, I was considering the "easier" implementation path.  Does that clear it up?

**Comment by joshsunshine** [comment]
@joshsunshine I understood that. I'm arguing that the "harder" implementation path should infer the `fillColor:?` from a program that doesn't have it and optimize the circle's fillColor and not the group's fillColor. Let me change your program slightly again:
```
x.circle = Circle {
   r: 100
   center: (0,0)
}
x.circle2 = Circle {
   r: 100
   center: (100,100)
}

x.group = Group {
   shapes: [ x.circle, x.circle2 ]
   fillColor: rgba( 0, 1, 0, 1 )
}

encourage equal( x.circle.fillColor, rgba( 1, 0, 0, 1 ) )
```
In this program, it would be very strange if both circles changed to red based on the objective. Instead only x.circle should be red.

**Comment by keenancrane** [comment]
>In this program, it would be very strange if both circles changed to red based on the objective.

Ok, I finally understand the question. :-)

It's definitely _not_ that strange in the case of transformations.  In a transformation hierarchy, all the ancestor transformations can and should be optimized in order to reach the objective of the child.  Yes, this may affect the transformation of objects other than the child.  But that's how inverse kinematics works!  Imagine a robot (or character, as illustrated above).  The person designing/describing the robot wants to be able to explicitly say how many degrees of freedom there are, e.g., how many joints get optimized.  Then solve for the best configuration that achieves the stated goals using these joints, and these joints only.  The system should not suddenly create new joints that the user didn't want in order to better achieve the goal.

I guess I feel the same way about other attributes as well.  If I only declared one color (in this case, the color of the group), then yes, I want different children to have to "compete" for control over the value of this color.  If two children want to be near two different colors, then the parent should try to negotiate between those colors (…just like a real parent??).  It shouldn't simply give each child exactly what it wants.  As in inverse kinematics, this tension is important, because it enables one to design diagrams that naturally balance or interpolate between different objectives.

**Comment by joshsunshine** [comment]
@keenancrane I'm glad we finally understand each other 😄 . I have no experience with inverse kinematics so I'll trust you that it's the way things are done. I know I'd be surprised, in the program above, if both circles turned red. Before this discussion I would also have been surprised if both circles moved based on an objective applied to one! I'm willing to suspend disbelief and go with your proposed "hard" semantics.  Hopefully this long discussion has been useful to whomever ultimately implements this feature.

**Comment by liangyiliang** [comment]
This is something I can work on! Transformations do remind me of forward kinematics ... I think I will get the basics working first before considering how transformations work.

Are there any particular reasons where a "name" for the group is necessary aside from editability? It seems that other shapes don't need `name`?
